common/transport/amqp/src/amqp.ts

```diff
@@ -132,6 +88,185 @@ export class Amqp {
+            The ‘to’ field of the message should be set to the ‘to’ argument.
+            The ‘body’ of the message should be built using the message argument.] */
+
+            let amqpMessage = AmqpMessage.fromMessage(message);
+            if (to !== undefined) {
+              amqpMessage.properties.to = to;
+            }
+
+            if (!this._senders[endpoint]) {
+              this._fsm.handle('attachSenderLink', endpoint, null, (err) => {
+                if (err) {
+                  done(err);
+                } else {
+                  (this._senders[endpoint] as SenderLink).send(amqpMessage, done);
+                }
+              });
+            } else {
+              (this._senders[endpoint] as SenderLink).send(amqpMessage, done);
+            }
+
+          },
+          getReceiver: (endpoint: string, done: GenericAmqpBaseCallback): void => {
+            /*Codes_SRS_NODE_COMMON_AMQP_16_010: [If a receiver for this endpoint doesn’t exist, the getReceiver method should create a new AmqpReceiver object and then call the done() method with the object that was just created as an argument.] */
+            if (!this._receivers[endpoint]) {
+              this._fsm.handle('attachReceiverLink', endpoint, null, (err) => {
+                if (err) {
+                  done(err);
+                } else {
+                  done(null, this._receivers[endpoint]);
+                }
+              });
+            } else {
+              /*Codes_SRS_NODE_COMMON_AMQP_16_009: [If a receiver for this endpoint has already been created, the getReceiver method should call the done() method with the existing instance as an argument.] */
+              this._safeCallback(done, null, this._receivers[endpoint]);
+            }
+          },
+          attachReceiverLink: (endpoint: string, linkOptions: any, done: GenericAmqpBaseCallback): void => {
+            let receiverFsm = new ReceiverLink(endpoint, linkOptions, this._amqp);
+            this._receivers[endpoint] = receiverFsm;
+            this._receivers[endpoint].on('error', (err) => {
+              debug('error on the receiver link for endpoint ' + endpoint);
+              debug(err.toString());
+              delete(this._receivers[endpoint]);
+            });
+            this._receivers[endpoint].attach(done);
+          },
+          attachSenderLink: (endpoint: string, linkOptions: any, done: GenericAmqpBaseCallback): void => {
+            let senderFsm = new SenderLink(endpoint, linkOptions, this._amqp);
+            this._senders[endpoint] = senderFsm;
+            this._senders[endpoint].on('error', (err) => {
+              debug('error on the sender link for endpoint ' + endpoint);
+              debug(err.toString());
+              delete(this._senders[endpoint]);
+            });
+            this._senders[endpoint].attach(done);
+          },
+          detachReceiverLink: (endpoint: string, detachCallback: GenericAmqpBaseCallback): void => {
```

Comment:
And this is why I approved.  I feel certain that this will be resolved in the fullness of time.

---
In reply to: [129160272](https://github.com/Azure/azure-iot-sdk-node/pull/55#discussion_r129160272) [](ancestors = 129160272,128132936)